mono/metadata/w32file.c

```diff
@@ -511,86 +511,91 @@ ves_icall_System_IO_MonoIO_FindCloseFile (HANDLE hnd)
 }
 
 /* FIXME make gc suspendable */
-MonoString *
-ves_icall_System_IO_MonoIO_FindFirst (MonoString *path,
-				      MonoString *path_with_pattern,
+MonoStringHandle
+ves_icall_System_IO_MonoIO_FindFirst (MonoStringHandle path,
+				      MonoStringHandle path_with_pattern,
 				      gint32 *result_attr, gint32 *ioerror,
 				      gpointer *handle)
 {
 	MonoError error;
 	WIN32_FIND_DATA data;
 	HANDLE find_handle;
 	IncrementalFind *ifh;
-	MonoString *result;
+	MonoStringHandle result = NULL_HANDLE_STRING;
 	
 	*ioerror = ERROR_SUCCESS;
 	
-	find_handle = mono_w32file_find_first (mono_string_chars (path_with_pattern), &data);
+	guint32 path_with_pattern_handle = 0;
+	const gunichar2 *const path_with_pattern_chars = mono_string_handle_pin_chars (path_with_pattern, &path_with_pattern_handle);
+	find_handle = mono_w32file_find_first (path_with_pattern_chars, &data);
+	mono_gchandle_free (path_with_pattern_handle);
 	
 	if (find_handle == INVALID_HANDLE_VALUE) {
 		gint32 find_error = mono_w32error_get_last ();
 		*handle = NULL;
 		
 		if (find_error == ERROR_FILE_NOT_FOUND) 
-			return NULL;
+			goto end;
 		
 		*ioerror = find_error;
-		return NULL;
+		goto end;
 	}
 
 	ifh = g_new (IncrementalFind, 1);
 	ifh->find_handle = find_handle;
-	ifh->utf8_path = mono_string_to_utf8_checked (path, &error);
+	ifh->utf8_path = mono_string_handle_to_utf8 (path, &error);
 	if (mono_error_set_pending_exception (&error)) {
```

Comment:
`HANDLES()` functions take a `MonoError *` argument so you don't need a local `&error`; they also don't need to call `mono_error_set_pending_exception` - the wrapper will do it.  So this should be `if (!is_ok (error))`